domain/secret/state/state.go

```diff
@@ -1010,6 +1011,80 @@ ON CONFLICT(revision_uuid, name) DO UPDATE SET
 	return nil
 }
 
+// InitialWatchStatementForOwnedSecrets returns the table name and query to use
+// for watching secret metadata changes.
+func (st State) InitialWatchStatementForOwnedSecrets(
+	appOwners domainsecret.ApplicationOwners, unitOwners domainsecret.UnitOwners,
+) (string, eventsource.NamespaceQuery) {
+	queryFunc := func(ctx context.Context, runner coredatabase.TxnRunner) ([]string, error) {
+		var ownedURIs []*coresecrets.URI
+		err := st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+			var err error
+			ownedURIs, err = st.GetSecretsForOwners(ctx, appOwners, unitOwners)
+			return errors.Trace(err)
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ids := make([]string, len(ownedURIs))
+		for i, uri := range ownedURIs {
+			ids[i] = uri.ID
+		}
+		return ids, nil
+	}
+	return "secret_metadata", queryFunc
+}
+
+// IsSecretOwnedBy returns whether the secret with the given URI is owned by the specified apps and/or units.
+func (st State) IsSecretOwnedBy(
+	ctx context.Context,
+	uri *coresecrets.URI,
+	appOwners domainsecret.ApplicationOwners, unitOwners domainsecret.UnitOwners,
+) (bool, error) {
+	if len(appOwners) == 0 && len(unitOwners) == 0 {
+		return false, errors.New("must supply at least one app owner or unit owner")
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	query := `
+SELECT COUNT(*) AS &count.num
```

Comment:
This SQL isn't quite right. We're missing a unit test for when the secret URL is not found, and adding one reveals the error: the below snippet fails because error is nil.
```
	uri3 := coresecrets.NewURI()
	_, err := st.IsSecretOwnedBy(ctx, uri3, domainsecret.ApplicationOwners{"mysql"}, nil)
	c.Check(err, jc.ErrorIs, secreterrors.SecretNotFound)
```

Reply:
Thanks for updating the comment to remove mention of the NotFound error.
Maybe just tweak the comment to mention that it's ok for the URI to not exist and that in that case false is returned, since in most/all other cases, the caller would expect a not found error.